File: packages/gavio-py/gavio/interceptors/audit/trace.py
"""Audit-chain verification (F-OBS-02) and multi-agent DAG trace (F-OBS-03)."""

from __future__ import annotations

from dataclasses import dataclass, field

from .record import AuditRecord


def verify_chain(records: list[AuditRecord]) -> bool:
    """Return True if the records form an intact hash chain.

    Each record's ``previous_hash`` must equal the SHA-256 content hash of the
    record before it. The first record's ``previous_hash`` must be empty.
    Any edit, reorder, or deletion breaks the chain.
    """
    prev_hash = ""
    for rec in records:
        if rec.previous_hash != prev_hash:
            return False
        prev_hash = rec.content_hash()
    return True


@dataclass
class TraceNode:
    """A node in the multi-agent call graph."""

    trace_id: str
    agent_id: str | None
    parent_trace_id: str | None
    children: list[TraceNode] = field(default_factory=list)


def build_call_graph(records: list[AuditRecord]) -> list[TraceNode]:
    """Reconstruct the multi-agent DAG from audit records.

    Uses ``parent_trace_id`` + ``trace_id`` to link calls. Returns the root
    nodes (those with no known parent). A single agent call yields one root.
    """
    nodes: dict[str, TraceNode] = {}
    for rec in records:
        nodes[rec.trace_id] = TraceNode(
            trace_id=rec.trace_id,
            agent_id=rec.agent_id,
            parent_trace_id=rec.parent_trace_id,
        )

    roots: list[TraceNode] = []
    for node in nodes.values():
        parent = nodes.get(node.parent_trace_id) if node.parent_trace_id else None
        if parent is not None:
            parent.children.append(node)
        else:
            roots.append(node)
    return roots

File: packages/gavio-js/src/interceptors/audit/trace.ts
/** Audit-chain verification (F-OBS-02) and multi-agent DAG trace (F-OBS-03). */

import type { AuditRecord } from './record.js'

/**
 * Return true if the records form an intact hash chain. Each record's
 * previousHash must equal the content hash of the record before it; the first
 * must be empty. Any edit, reorder, or deletion breaks the chain.
 */
export function verifyChain(records: AuditRecord[]): boolean {
  let prevHash = ''
  for (const rec of records) {
    if (rec.previousHash !== prevHash) return false
    prevHash = rec.contentHash()
  }
  return true
}

export interface TraceNode {
  traceId: string
  agentId: string | null
  parentTraceId: string | null
  children: TraceNode[]
}

/**
 * Reconstruct the multi-agent DAG from audit records using parentTraceId +
 * traceId. Returns the root nodes (those with no known parent).
 */
export function buildCallGraph(records: AuditRecord[]): TraceNode[] {
  const nodes = new Map<string, TraceNode>()
  for (const rec of records) {
    nodes.set(rec.traceId, {
      traceId: rec.traceId,
      agentId: rec.agentId,
      parentTraceId: rec.parentTraceId,
      children: [],
    })
  }
  const roots: TraceNode[] = []
  for (const node of nodes.values()) {
    const parent = node.parentTraceId ? nodes.get(node.parentTraceId) : undefined
    if (parent) parent.children.push(node)
    else roots.push(node)
  }
  return roots
}
